### PR TITLE
Add rule: [a11y] no-aria-attributes-on-unsupported-elements

### PR DESCRIPTION
**Pull request by lifeart**
refs: https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/aria-unsupported-elements.md
https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_aria_12

### CODE REVIEW

File: test/unit/rules/lint-aria-unsupported-elements-test.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+const generateRuleTests = require('../../helpers/rule-test-harness');
+
+generateRuleTests({
+  name: 'aria-unsupported-elements',
+
+  config: true,
+
+  good: ['<div role="main-navigation"></div>', '<img aria-role="some-role">'],
```
Comment (MelSumner) [inline test/unit/rules/lint-aria-unsupported-elements-test.js]: these good examples are both incorrect.

File: docs/rule/aria-unsupported-elements.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+## aria-unsupported-elements
```
Comment (rwjblue) [inline docs/rule/aria-unsupported-elements.md]:
```suggestion
## no-aria-attributes-on-unsupported-elements
```

File: test/unit/rules/lint-aria-unsupported-elements-test.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
+const generateRuleTests = require('../../helpers/rule-test-harness');
+
+generateRuleTests({
+  name: 'aria-unsupported-elements',
+
+  config: true,
+
+  good: ['<div role="button"></div>', '<img role="option">'],
```
Comment (rwjblue) [inline test/unit/rules/lint-aria-unsupported-elements-test.js]: Can you add `good` examples using the prohibited elements also?

File: test/unit/rules/lint-aria-unsupported-elements-test.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+  bad: [
+    {
+      template: '<meta role>',
```
Comment (rwjblue) [inline test/unit/rules/lint-aria-unsupported-elements-test.js]: Lets make the tests as representative as possible, I don't think `role` attribute without a value is a valid thing, is it?

Comment (lifeart) [reply test/unit/rules/lint-aria-unsupported-elements-test.js]: fixed

File: test/unit/rules/lint-aria-unsupported-elements-test.js (hunk 3 above)
Comment (lifeart) [reply test/unit/rules/lint-aria-unsupported-elements-test.js]: done

File: docs/rule/aria-unsupported-elements.md (hunk 2 above)
Comment (lifeart) [reply docs/rule/aria-unsupported-elements.md]: done

File: docs/rule/no-aria-attributes-on-unsupported-elements.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+This rule **allows** the following:
+
+```hbs
+<meta>
+```
+
+### References
+
+ - Google Audit defs [AX_ARIA_12](https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-rules#ax_aria_12)
+ - [eslint-plugin-jsx-a11y/aria-unsupported-elements](https://github.com/evcohen/eslint-plugin-jsx-a11y/blob/master/docs/rules/aria-unsupported-elements.md#aria-unsupported-elements)
```
Comment (MelSumner) [inline docs/rule/no-aria-attributes-on-unsupported-elements.md]: please remove this line- we must not reference other people's rules, but rather the specification itself.

File: test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+  good: [
+    '<div role="button"></div>',
+    '<img role="option">',
+    '<meta data-name="foo">',
+    '<script src="noop"></script>',
+    '<html title="noop"></html>',
+  ],
+
+  bad: [
```
Comment (MelSumner) [inline test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js]: After looking at these "bad" tests, I think we could abstract them a bit more, to consolidate them.

**Comment by MelSumner** [comment]
> I wonder if we should instead have a `no-invalid-role` rule? It could check for `role` on invalid elements (which is what this PR currently does), but it could also check to ensure that the `role` value is within the valid list of values.
> 
> @MelSumner - What do you think?

I wonder if that might be a job for ember language server? let me look to see how complex that would get.

File: docs/rule/no-aria-attributes-on-unsupported-elements.md (hunk 5 above)
Comment (MelSumner) [inline docs/rule/no-aria-attributes-on-unsupported-elements.md:27]:
```suggestion
 - [WCAG 4.1.2 - Name, Role, Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value)
```

File: test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+        source: '<meta role="button">',
+        line: 1,
+        column: 0,
+      },
+    },
+    {
+      template: '<meta aria-foo="some-prop">',
+
+      result: {
+        message: 'meta tag must not contain "aria-foo" attribute',
```
Comment (MelSumner) [inline test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js:32]:
```suggestion
        message: 'The meta tag must not contain an "aria-*" attribute',
```

File: test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        source: '<meta aria-foo="some-prop">',
+        line: 1,
+        column: 0,
+      },
+    },
+    {
+      template: '<meta role="name">',
+
+      result: {
+        message: 'meta tag must not contain "role" attribute',
```
Comment (MelSumner) [inline test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js:43]:
```suggestion
        message: 'The meta tag must not contain the "role" attribute',
```

File: test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+        source: '<meta role="name">',
+        line: 1,
+        column: 0,
+      },
+    },
+    {
+      template: '<html role="button"></html>',
+
+      result: {
+        message: 'html tag must not contain "role" attribute',
```
Comment (MelSumner) [inline test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js:54]:
```suggestion
        message: 'The html tag must not contain the "role" attribute',
```

File: test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+        source: '<html role="button"></html>',
+        line: 1,
+        column: 0,
+      },
+    },
+    {
+      template: '<script role="template"></script>',
+
+      result: {
+        message: 'script tag must not contain "role" attribute',
```
Comment (MelSumner) [inline test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js:65]:
```suggestion
        message: 'A script tag must not contain the "role" attribute',
```

File: test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        source: '<script role="template"></script>',
+        line: 1,
+        column: 0,
+      },
+    },
+    {
+      template: '<style role="style"></style>',
+
+      result: {
+        message: 'style tag must not contain "role" attribute',
```
Comment (MelSumner) [inline test/unit/rules/lint-no-aria-attributes-on-unsupported-elements-test.js:76]:
```suggestion
        message: 'A style tag must not contain the "role" attribute',
```

**Pull request closed without merging**